### PR TITLE
[Merged by Bors] - feat: add definition of opposite of subsemirings, subrings, and subalgebras

### PR DESCRIPTION
**Pull request by acmepjz**
similar to that in `Mathlib.Algebra.Group.Subgroup.MulOpposite`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

discussion: <https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Subalgebra.2EequivOpposite>

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you copy the `Submonoid.MulOpposite` file here instead, and build up separate `op` and `unop` definitions?

**Comment by acmepjz** [comment]
> Can you copy the `Submonoid.MulOpposite` file here instead, and build up separate `op` and `unop` definitions?

Let me check later.

**Comment by acmepjz** [comment]
> Can you copy the `Submonoid.MulOpposite` file here instead, and build up separate `op` and `unop` definitions?

Done.

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+variable {ι R A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
+
+/-- Pull a subalgebra back to an opposite subalgebra along `MulOpposite.unop` -/
+@[simps toSubsemiring]
+protected def op (S : Subalgebra R A) : Subalgebra R Aᵐᵒᵖ where
+  toSubsemiring := S.toSubsemiring.op
+  algebraMap_mem' := S.algebraMap_mem
+
+@[simp]
```
Comment (YaelDillies) [inline Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:30]:
```suggestion
@[simp, norm_cast]
```
and everywhere else

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  MulOpposite.unop_surjective.forall
+
+/-- A subalgebra `S` of `A / R` determines a subring `S.op` of the opposite ring `Aᵐᵒᵖ / R`. -/
+@[simps]
+def opEquiv : Subalgebra R A ≃o Subalgebra R Aᵐᵒᵖ where
+  toFun := Subalgebra.op
+  invFun := Subalgebra.unop
+  left_inv := unop_op
+  right_inv := op_unop
+  map_rel_iff' := op_le_op_iff
```
Comment (YaelDillies) [inline Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:77]: We don't have this for `Subgroup`. Can you add it?

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+theorem op_le_op_iff {S₁ S₂ : Subalgebra R A} : S₁.op ≤ S₂.op ↔ S₁ ≤ S₂ :=
+  MulOpposite.op_surjective.forall
+
+@[simp]
+theorem unop_le_unop_iff {S₁ S₂ : Subalgebra R Aᵐᵒᵖ} : S₁.unop ≤ S₂.unop ↔ S₁ ≤ S₂ :=
+  MulOpposite.unop_surjective.forall
+
+/-- A subalgebra `S` of `A / R` determines a subring `S.op` of the opposite ring `Aᵐᵒᵖ / R`. -/
+@[simps]
+def opEquiv : Subalgebra R A ≃o Subalgebra R Aᵐᵒᵖ where
```
Comment (YaelDillies) [inline Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:72]:
```suggestion
def opOrderIso : Subalgebra R A ≃o Subalgebra R Aᵐᵒᵖ where
```

Comment (acmepjz) [reply Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:72]: But it's called [Submonoid.opEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/MulOpposite.html#Submonoid.opEquiv) for Submonoid, etc.

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean (hunk 2 above)
Comment (acmepjz) [reply Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:77]: We already have [Subgroup.opEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Subgroup/MulOpposite.html#Subgroup.opEquiv) unless I mistake something.

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:72]: Uh, okay 😦

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean:77]: Indeed, I missed it

File: Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean (hunk 1 above)
Comment (acmepjz) [reply Mathlib/Algebra/Algebra/Subalgebra/MulOpposite.lean]: Done.

**Comment by YaelDillies** [comment]
Thanks!

maintainer merge

**Review by eric-wieser — APPROVED** [review]
Just the `ι` issue to solve (in all three files)

File: Mathlib/Algebra/Ring/Subsemiring/MulOpposite.lean [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+# Subsemiring of opposite semirings
+
+For every semiring `R`, we construct an equivalence between subsemirings of `R` and that of `Rᵐᵒᵖ`.
+
+-/
+
+namespace Subsemiring
+
+variable {ι R : Type*} [NonAssocSemiring R]
```
[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/Algebra/Ring/Subsemiring/MulOpposite.lean]: Done.

**Review by eric-wieser — APPROVED** [review]
bors d+

thanks!

**Pull request closed without merging**